lxd/project/permissions.go

```diff
@@ -1514,3 +1514,115 @@ func AllowSnapshotCreation(p *api.Project) error {
 
 	return nil
 }
+
+// GetRestrictedClusterGroups returns a slice of restricted cluster groups for the given project.
+func GetRestrictedClusterGroups(p *api.Project) []string {
+	return shared.SplitNTrimSpace(p.Config["restricted.cluster.groups"], ",", -1, true)
+}
+
+// AllowClusterMember returns nil if the given project is allowed to use the cluster member.
+func AllowClusterMember(p *api.Project, member *db.NodeInfo) error {
+	clusterGroupsAllowed := GetRestrictedClusterGroups(p)
+
+	if shared.IsTrue(p.Config["restricted"]) && len(clusterGroupsAllowed) > 0 {
+		found := false
+		for _, memberGroupName := range member.Groups {
+			if shared.StringInSlice(memberGroupName, clusterGroupsAllowed) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return fmt.Errorf("Project isn't allowed to use this cluster member")
+		}
+	}
+
+	return nil
+}
+
+// AllowClusterGroup returns nil if the given project is allowed to use the cluster group.
+func AllowClusterGroup(p *api.Project, group string) error {
+	clusterGroupsAllowed := GetRestrictedClusterGroups(p)
+
+	// Skip the check if the project is not restricted
+	if shared.IsFalseOrEmpty(p.Config["restricted"]) {
+		return nil
+	}
+
+	if len(clusterGroupsAllowed) > 0 && !shared.StringInSlice(group, clusterGroupsAllowed) {
+		return fmt.Errorf("Project isn't allowed to use this cluster group")
+	}
+
+	return nil
+}
+
+// CheckTargetMember checks if the given member is allowed for the project.
+// If nil it also returns the resolved target member node information.
+func CheckTargetMember(tx *db.ClusterTx, p *api.Project, targetMember string) (*db.NodeInfo, error) {
+	ctx := context.Background()
```

Comment:
You should accept a ctx argument instead and leave it up to the caller what they provide.

Reply:
Done for the entire call chain.